### PR TITLE
plugin-testng: Plugin initial commit

### PR DESCRIPTION
**Pull request by vanisrikanithi**
Plugin-TestNG: This PR includes the Implementation of TestNG plugin to support processing the TestNG XML reports.

For Harness CI Pipeline test result please click on the link [here](https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/module/ci/orgs/default/projects/jenkinstoharness/pipelines/Vani_TestNG/deployments/c2C2V9LmSpG9ysjN4xOB8w/pipeline?storeType=INLINE)

For Harness Test Release Pipeline of Testng plugin click [here](https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/module/ci/orgs/default/projects/jenkinstoharness/pipelines/VaniTestNGdronereleasepipeline/deployments/qLRb8gTgTOy56PP3xy3qBQ/pipeline?storeType=INLINE&stage=JMdv8BloTPaTtEaHezQP1A&step=7dWREgVQRSeBlFXrQ08pDg&childStage=&stageExecId=)

Harness Pipeline PLUGIN_FAILURE_ON_FAILED_TEST_CONFIG=true click [Here](https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/module/ci/orgs/default/projects/jenkinstoharness/pipelines/Vani_TestNG/deployments/lBROuA0qTqevDYpqlIcMvw/pipeline)

@Ompragash Please review the PR for TestNG plugin and add your comments if there are any.

### CODE REVIEW

File: plugin/plugin.go [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+		logger := logrus.WithError(err)
+		logger.Error("Error locating files")
+		return errors.New("failed to locate files: " + err.Error())
+	}
+
+	if len(files) == 0 {
+		if args.PluginFailIfNoResults {
+			return errors.New("no TestNG XML report files found. Check the report file pattern")
+		}
+		logrus.Warn("No TestNG XML report files found, continuing execution as PluginFailIfNoResults is false")
```
Comment (Ompragash) [inline plugin/plugin.go]: lets remove PluginFailIfNoResults and just throw error when the report file isn't available

File: plugin/plugin.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+)
+
+// Args represents the plugin's configurable arguments.
+type Args struct {
+	ReportFilenamePattern     string `envconfig:"PLUGIN_REPORT_FILENAME_PATTERN"`
+	FailedFails               int    `envconfig:"PLUGIN_FAILED_FAILS"`
+	FailedSkips               int    `envconfig:"PLUGIN_FAILED_SKIPS"`
+	FailureOnFailedTestConfig bool   `envconfig:"PLUGIN_FAILURE_ON_FAILED_TEST_CONFIG"`
+	UnstableFails             int    `envconfig:"PLUGIN_UNSTABLE_FAILS"`
+	UnstableSkips             int    `envconfig:"PLUGIN_UNSTABLE_SKIPS"`
```
Comment (Ompragash) [inline plugin/plugin.go]: let's remove inputs related to unstable as Harness doesn't support the pipeline status as unstable

File: plugin/plugin_test.go
Comment (Ompragash) [inline plugin/plugin_test.go:1]: remove tests covering unstable and also add tests for below:
- incorrect XML structure
- mixed valid and invalid files
- edge case: validate plugin behaviour with very high values or empty results

File: README.md [hunk 3]
```diff
@@ -1 +1,91 @@
...
+Build the plugin image:
+
+```text
+docker build -t plugins/testng -f docker/Dockerfile .
+```
+
+## Testing
+
+Execute the plugin from your current working directory:
+## TestNG: This plugin converts TestNG XML reports (generated using org.testng.reporters.XMLReporter) to JUnit-compatible XML reports, enforcing thresholds and configuration validations to manage build statuses.
```
Comment (Ompragash) [inline README.md]: please remove this and the references of Junit if any

**Review by Ompragash — CHANGES_REQUESTED** [review]

File: plugin/plugin.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+// ValidateInputs ensures the user inputs meet the plugin requirements.
+func ValidateInputs(args Args) error {
+	if args.ReportFilenamePattern == "" {
+		return errors.New("missing required parameter: ReportFilenamePattern. Please specify the pattern to locate the TestNG report files")
+	}
+	if args.FailedFails < 0 || args.FailedSkips < 0 || args.UnstableFails < 0 || args.UnstableSkips < 0 {
+		return errors.New("threshold values must be non-negative. Check the configured values for failed and skipped tests")
+	}
+	if args.ThresholdMode != 1 && args.ThresholdMode != 2 {
```
Comment (Ompragash) [inline plugin/plugin.go]: instead of 1 and 2, users can directly input `absolute` or `percentage` which are self-explanatory and also set the below constants with a default thresholdmode and update the switch case in `validateThresholds` accordingly 
```go
const (
    ThresholdModeAbsolute   = "absolute"
    ThresholdModePercentage = "percentage"
    DefaultThresholdMode    = ThresholdModeAbsolute // Default value
)
```

with the above default, when no PLUGIN_THRESHOLD_MODE set by the user, mode `absolute` will be used and logs the below
```
PLUGIN_THRESHOLD_MODE not specified. Defaulting to 'absolute'
```

update the same in the readme as well

File: plugin/plugin.go [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+		})
+		logger.Error(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+// locateFiles identifies files matching the given pattern.
+func locateFiles(pattern string) ([]string, error) {
```
Comment (Ompragash) [inline plugin/plugin.go:126]: add checks for file read permissions to prevent runtime errors and log the number of files found for clarity

File: plugin/plugin.go [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+		return nil, errors.New("failed to search for files: " + err.Error())
+	}
+	if len(matches) == 0 {
+		return nil, errors.New("no files found matching the report filename pattern")
+	}
+	return matches, nil
+}
+
+// processFile reads a TestNG XML report and logs details.
+func processFile(filename string) (Results, error) {
```
Comment (Ompragash) [inline plugin/plugin.go:167]: add logic to handle permission issues and log errors. make sure the logs are clear between `file not found` & `parsing error` errors

File: plugin/plugin.go [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+		logger := logrus.WithError(err).WithField("File", filename)
+		logger.Error("Failed to read file")
+		return Results{}, errors.New("failed to read file: " + err.Error())
+	}
+
+	var report TestNGReport
+	if err := xml.Unmarshal(data, &report); err != nil {
+		logger := logrus.WithError(err).WithField("File", filename)
+		logger.Error("Failed to parse TestNG XML")
+		return Results{}, errors.New("failed to parse TestNG XML: " + err.Error())
```
Comment (Ompragash) [inline plugin/plugin.go]: - validate essential fields like Suites and Classes are present and non-empty, also warn about structural issues in the XML content, like missing test classes or empty suites like "File testng-report.xml contains no test suites in the XML structure"

- also in case of multiple reports parsing, if parsing one report fails, add logic to skip the file and continue processing other reports

File: plugin/plugin.go [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+		return fmt.Errorf("\nnumber of skipped tests (%d) exceeded the skip threshold (%d)", results.Skipped, args.FailedSkips)
+	}
+	return nil
+}
+
+// validatePercentageThresholds checks percentage-based thresholds.
+func validatePercentageThresholds(results Results, args Args) error {
+	totalTests := results.Total
+	if totalTests == 0 {
+		return nil // No tests to validate
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
	if totalTests == 0 {
	        logrus.Warn("No tests executed; skipping percentage-based threshold validation.")
		return nil
```

also add a test case to verify behavior when no tests are executed

File: plugin/plugin.go [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+			}
+		}
+	default:
+		return fmt.Errorf("\ninvalid ThresholdMode: %d, expected 1 (absolute) or 2 (percentage)", args.ThresholdMode)
+	}
+	return nil
+}
+
+// validateAbsoluteThresholds checks absolute thresholds.
+func validateAbsoluteThresholds(results Results, args Args) error {
```
Comment (Ompragash) [inline plugin/plugin.go]: move the repetitive checks into helper functions and make use of them in both validateAbsoluteThresholds and validatePercentageThresholds

File: plugin/plugin.go [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+		logger := logrus.WithError(err).WithField("File", filename)
+		logger.Error("Failed to parse TestNG XML")
+		return Results{}, errors.New("failed to parse TestNG XML: " + err.Error())
+	}
+
+	return logTestNGReportDetails(report), nil
+}
+
+// logTestNGReportDetails logs the details of a TestNG report and returns the aggregated results.
+func logTestNGReportDetails(report TestNGReport) Results {
```
Comment (Ompragash) [inline plugin/plugin.go:212]: add check to handle invalid durations by adding a warning log for invalid or missing DurationMS values and also log suite-level summaries 

again create helper functions to reduce duplication in aggregation logic

once the above completed, add uts for aggregation logic

File: plugin/plugin.go [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+		return nil, errors.New("no files found matching the report filename pattern")
+	}
+	return matches, nil
+}
+
+// processFile reads a TestNG XML report and logs details.
+func processFile(filename string) (Results, error) {
+	logrus.Infof("Processing file: %s", filename)
+
+	data, err := os.ReadFile(filename)
```
Comment (Ompragash) [inline plugin/plugin.go]: instead of loading the entire file into memory using os.ReadFile use xml.Decoder for streaming to avoid memory spike and test the behaviour

**Review by Ompragash — CHANGES_REQUESTED** [review]

File: plugin/plugin.go [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+		if err != nil {
+			logrus.WithField("File", file).WithError(err).Warn("Skipping file due to errors")
+			skippedFiles = append(skippedFiles, file)
+			continue // Skip this file and process others
+		}
+		aggregatedResults.Total += results.Total
+		aggregatedResults.Failures += results.Failures
+		aggregatedResults.Skipped += results.Skipped
+		aggregatedResults.DurationMS += results.DurationMS
+	}
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
        var (
                resultsChan    = make(chan Results, len(files))
                errorsChan     = make(chan error, len(files))
        )
        
        for _, file := range files {
	        go func(f string) {
		        res, err := processFile(f)
		        if err != nil {
			        errorsChan <- fmt.Errorf("failed to process file %s: %w", f, err)
			        return
		        }
		        resultsChan <- res
	        }(file)
        }
        
	var aggregatedResults Results
	var skippedFiles []string
	
	for i := 0; i < len(files); i++ {
	        select {
	        case res := <-resultsChan:
		        aggregatedResults.Total += res.Total
		        aggregatedResults.Failures += res.Failures
		        aggregatedResults.Skipped += res.Skipped
		        aggregatedResults.DurationMS += res.DurationMS
	        case err := <-errorsChan:
		        logrus.Warn(err)
			if e, ok := err.(*os.PathError); ok {
			        skippedFiles = append(skippedFiles, e.Path)
			}
	        }
	}
```

let's implement parallel processing in case of large number of files

Comment (Ompragash) [reply plugin/plugin.go]: also add tests with a mix of valid and invalid files in a parallel processing scenario

**Review by Ompragash — CHANGES_REQUESTED** [review]
after addressing the changes, build and push the images for both Linux and Windows and share the test execution link

**Comment by vanisrikanithi** [comment]
@Ompragash  added parallel process code and tested successfully, could you please check it and confirm..

File: plugin/plugin.go [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+		logSuiteGroups(suite)
+		logSuiteTestDetails(suite)
+	}
+
+	// Log aggregated results with failed and skipped test names
+	if len(failedTests) > 0 {
+		logrus.Infof("\nTest case Failures: %s", formatTestNames(failedTests))
+	}
+	if len(skippedTests) > 0 {
+		logrus.Infof("\nTest case Skips: %s", formatTestNames(skippedTests))
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
		logrus.Infof("\nFailed Test cases: %s", formatTestNames(failedTests))
	}
	if len(skippedTests) > 0 {
		logrus.Infof("\nSkipped Test cases: %s", formatTestNames(skippedTests))
```
Screenshot 2024-12-17 at 8 56 54 PM
add `======` before `Failed Test cases` for clarity

**Review by Ompragash — APPROVED** [review]
